cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+structure comprising one or more fields. These fields are stored in structure-of-arrays format, so 
+that the column of structs has a nested column for each field of the structure. 
+
+As the heading implies, these column types may be nested arbitrarily. One may create a column of 
+lists of structs, where the fields of the struct may be of any type, including strings, lists and 
+structs. Thinking about deeply nested data types can be confusing for column-based data, even with 
+experience. Therefore it is important to carefully write algorithms, and to test and document them
+well.
+
+## List columns
+
+In order to represent variable-width elements, libcudf columns contain a vector of child columns.
+For list columns, the parent column's type is `LIST` and contains no data, but its size represents
+the number of lists in the column, and its null mask represents the validity of each list element.
+The parent has two children. 
+
+1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each list
+   in a dense column of elements.
+2. A column containing the actual data and optional null mask for all elements of all the lists 
+   packed together.
+   
+With this representation, `data[offsets[i]]` is the first element of list `i`, and the size of list
+`i` is given by `offsets[i+1] - offsets[i]`.
+
+Note that the data may be of any type, and therefore the data column may itself be a nested column
+of any type. Note also that not only is each list nullable (using the null mask of the parent), but
+each list element may be nullable. So you may have a lists column with null row 3, and also null
+element 2 of row 4.
+
+## Strings columns
+
+Strings are represented in much the same way as lists, except that the data child column is always 
+a non-nullable column of `INT8` data. The parent column's type is `STRING` and contains no data,
+but its size represents the number of strings in the column, and its null mask represents the 
+validity of each string. To summarize, the strings column children are:
+
+1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each 
+   string in a dense column of all characters.
+2. A non-nullable column of `INT8` elements of all the characters across all the strings packed 
+   together.
+
+With this representation, `characters[offsets[i]]` is the first character of string `i`, and the 
+size of string `i` is given by `offsets[i+1] - offsets[i]`. The following image shows an example of
+this compound column representation of strings.
+
+![strings](strings.png)
+
+## Structs columns
+
+Structs are represented similarly to lists, except that they have multiple child data columns.
+The parent column's type is `STRUCT` and contains no data, but its size represents the number of 
+structs in the column, and its null mask represents the validity of each struct element. The parent 
```

Comment:
I'm trying to decide whether the interplay between the validity masks of structs and its members should be described here. Specifically that the parent's validity mask is bitwise-ANDed with all its children's masks, so that a null `struct` row at the parent level marks the corresponding child row to null, for all child columns.

Would an explanation (with examples)  make this section too verbose?

Reply:
It might, but it's better to be too verbose and solve that problem by refactoring than to have missing information. If you can provide something in a suggestion comment, I'll accept it.

Reply:
I'm not sure I would get this written correctly by myself. Would be great if you could write it, either in this PR or file an issue after and do it as a followup.

Reply:
Sorry I missed this. I'll add a section shortly.

Reply:
Hmm... I can't seem to add a `suggestion` to this diff. 
Perhaps something like this? (Will likely need rephrasing.)
Please note that the "\\```" will need the "\\" removed. I couldn't figure out how to escape this correctly in the suggestion text. :/

```
As per the [Apache Arrow specification](https://arrow.apache.org/docs/format/Columnar.html#struct-layout), in addition to the struct column's null mask, each struct
field column has its own optional null mask. A struct field's validity can vary independently from the 
corresponding struct row. For instance, a non-null struct row might have a null field. However, the fields
of a null struct row are deemed to be null as well.
As an example, consider a struct column of type `STRUCT<FLOAT32, INT32>`. If the contents were
`[ {1.0, 2}, {4.0, 5}, null, {8.0, null} ]`, the struct column's layout would be as follows:

\```
{
  type = STRUCT
  null_mask = 1011
  null_count = 1
  children = {
    {   
      type = FLOAT32
      data = [1.0, 4.0, X, 8.0]
      null_mask  = 1011
      null_count = 1
    },  
    {   
      type = INT32
      data = [2, 5, X, X]
      null_mask  = 0011
      null_count = 2
    }  
  }   
}
\```
The last struct row (index 3) is not null, but has a null value for its INT32 field. This is 
reflected in bit 3 of the null masks of the struct column and the INT32 field.
Also, row 2 of the struct column is null. Its corresponding fields should also be null, so bit 2 is
unset in the null masks of both struct fields.

```

Reply:
I believe the bit indices are shown right-to-left `76543210`. 
So `1011` means that only `row0`, `row1`, and `row3` are valid.

Reply:
1 == valid. 0 == null.
I based the null-mask representation on how it shows up when the column is printed. I realize that this is confusing.

I have modified the example to show the null_mask as an array instead. This should be clearer.

Reply:
@davidwendt, et al, 

Would anyone know how one makes a "```suggestion" block that itself contains a "```" code block, in GitHub? I'd like to make my suggestion above available to pick into this PR, but I'm thwarted by GitHub's formatting. :/

Reply:
It looks like you may be able to change the 3 backtick marks to 3 tildes for the suggestion block.
`````
~~~suggestion
Some text.
```
Example text.
```
Other text.
~~~
`````

I think it is also possible to nest blocks by using more than 3 backticks.

Reply:
@mythrocks the problem I have with your example is this: 

> The last struct row (index 3) is not null, but has a null value in the INT32 field. This is reflected in bit 3 of the null masks of the struct column and the INT32 field.

But the column-level (parent-level?) null mask is [1, 1, 0, 1] (bit 3 is valid), while the INT32 field null mask is [1, 1, 0, 0] (bit 3 is invalid/null).  So your text does not match your example.

Reply:
>> The last struct row (index 3) is not null...

> But the column-level (parent-level?) null mask is [1, 1, 0, 1] (bit 3 is valid)

A) `1` indicates valid. In the `STRUCT` mask, bit 3 is 1. Therefore, `STRUCT` at index 3 is valid, i.e. not null. The two quotes above are not at odds.

>> but has a null value in the INT32 field...

> while the INT32 field null mask is [1, 1, 0, 0] (bit 3 is invalid/null).

B) So the `INT32` field's mask[3] is 0, i.e. invalid.

From A and B above, the `STRUCT` is valid, but its `INT32` field is invalid. Hence, `{8.0, ∅}`.

Forgive me, I'm still unable to see the problem. :/

Reply:
Sorry, I read 

> This is reflected in bit 3 of the null masks of the struct column and the INT32 field.

To mean that the null masks of the struct column *and* the INT32 field should be null at bit 3. I will rewrite this sentence to be clear, it's unclear currently.

Reply:
Apologies for the clumsy wording. I meant that "this" is reflected in bit 3 of the null masks of the `STRUCT` column and the `INT32` field (i.e. `1` and `0` respectively).